### PR TITLE
Fix export of arange with data-dependent float step

### LINKED ISSUE #153247: Cannot use torch.arange in torch.ops.higher_order.scan
### 🐛 Describe the bug

I can't find any guard (``torch._check(...)``) to make that example work, the first one to fail is this one: https://github.com/pytorch/pytorch/blob/main/torch/_prims_common/__init__.py#L271. But that's probably not the only one.

```python

import torch

def filter_position_ids(
    patch_attention_mask: torch.Tensor,
    position_ids: torch.Tensor,
    boundaries: torch.Tensor,
    num_patches_per_side: int,
):
    for batch_idx, p_attn_mask in enumerate(patch_attention_mask):
        fractional_coords_h = torch.arange(0, 1 - 1e-6, 1 / p_attn_mask[:, 0].sum())
        fractional_coords_w = torch.arange(0, 1 - 1e-6, 1 / p_attn_mask[0].sum())

        bucket_coords_h = torch.bucketize(fractional_coords_h, boundaries, right=True)
        bucket_coords_w = torch.bucketize(fractional_coords_w, boundaries, right=True)

        pos_ids = (
            bucket_coords_h[:, None] * num_patches_per_side + bucket_coords_w
        ).flatten()
        position_ids[batch_idx][p_attn_mask.view(-1)] = pos_ids
    return position_ids

def float_arange(start, end, step):
    length = torch.sym_int((end - start) / step + (step * (1 - 1e-6)))
    torch._check(length > 0)
    res = torch.arange(0, length)
    torch._check(res.is_contiguous())
    fres = res.to(torch.float32)
    fstart = torch.tensor(start, dtype=torch.float32)
    return fres + fstart

def scan_filter_position_ids(
    patch_attention_mask: torch.Tensor,
    position_ids: torch.Tensor,
    boundaries: torch.Tensor,
    num_patches_per_side: int,
):

    def body(p_attn_mask, position_ids_row):
        fractional_coords_h = torch.arange(0, 1 - 1e-6, 1 / p_attn_mask[:, 0].sum().item())
        fractional_coords_w = torch.arange(0, 1 - 1e-6, 1 / p_attn_mask[0].sum().item())

        bucket_coords_h = torch.bucketize(fractional_coords_h, boundaries, right=True)
        bucket_coords_w = torch.bucketize(fractional_coords_w, boundaries, right=True)

        pos_ids = (
            bucket_coords_h[:, None] * num_patches_per_side + bucket_coords_w
        ).flatten()

        row = position_ids_row.clone()
        row[p_attn_mask.view(-1)] = pos_ids
        return [row]

    return torch.ops.higher_order.scan(
        body, [], [patch_attention_mask, position_ids], additional_inputs=[]
    )

class Model(torch.nn.Module):
    def forward(self, patch_attention_mask, position_ids, boundaries):
        if is_torchdynamo_exporting():
            res = scan_filter_position_ids(
                patch_attention_mask, position_ids, boundaries, 32
            )
            return res[0]
        return filter_position_ids(patch_attention_mask, position_ids, boundaries, 32)

patch_attention_mask = torch.randint(0, 20, (32, 32, 32)) >= 1
patch_attention_mask[:, :, :] = True
position_ids = torch.zeros((32, 1024), dtype=torch.int64)
boundaries = (torch.arange(33).to(torch.float32) / 33)[1:-1]
inputs = (patch_attention_mask, position_ids, boundaries)
model = Model()
expected = model(*inputs)
with fake_torchdynamo_exporting():
    got = model(*inputs)
torch.testing.assert_close(expected, got)

DYN = torch.export.Dim.DYNAMIC
ep = torch.export.export(model, inputs, dynamic_shapes=({0: DYN}, {0: DYN}, {0: DYN}))
torch.testing.assert_close(expected, ep.module()(*inputs))
```

### Versions

```
Collecting environment information...
PyTorch version: 2.8.0.dev20250506+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.35

Python version: 3.12.9 (main, Feb  5 2025, 08:49:00) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.6.68
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Laptop GPU
Nvidia driver version: 538.92
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.3.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               20
On-line CPU(s) list:                  0-19
Vendor ID:                            GenuineIntel
Model name:                           13th Gen Intel(R) Core(TM) i7-13800H
CPU family:                           6
Model:                                186
Thread(s) per core:                   2
Core(s) per socket:                   10
Socket(s):                            1
Stepping:                             2
BogoMIPS:                             5836.73
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves avx_vnni umip waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize flush_l1d arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            480 KiB (10 instances)
L1i cache:                            320 KiB (10 instances)
L2 cache:                             12.5 MiB (10 instances)
L3 cache:                             24 MiB (1 instance)
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] model-explorer-onnx==0.3.4
[pip3] mypy==1.15.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.2.5
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnx==1.19.0
[pip3] onnx-array-api==0.3.0
[pip3] onnx-extended==0.4.0
[pip3] onnxruntime_extensions==0.13.0
[pip3] onnxruntime-genai-cuda==0.6.0
[pip3] onnxruntime-training==1.23.0+cu126
[pip3] onnxscript==0.3.0.dev20250301
[pip3] optree==0.14.1
[pip3] pytorch-triton==3.3.0+git96316ce5
[pip3] torch==2.8.0.dev20250506+cu126
[pip3] torch_geometric==2.4.0
[pip3] torchaudio==2.6.0.dev20250506+cu126
[pip3] torchmetrics==1.6.2
[pip3] torchvision==0.22.0.dev20250506+cu126
[pip3] triton==3.2.0
[conda] Could not collect
```

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185568

`torch.arange` can receive a step derived from a captured scalar, such as
`1 / x.sum()` or `1 / mask.sum().item()`. During export/fake execution this
creates unbacked symbolic scalar values. The reference `arange` implementation
was using Python control flow on the step sign, which forced a guard on the
unbacked SymFloat before export had a bindable value for it.

Use `guard_or_false` for the sign-dependent consistency checks so metadata
execution does not branch on data-dependent symbolic values. In the scan path,
ignore fresh unbacked symbols only while evaluating the body for fake metadata,
matching the fact that those body-local symbols are represented by the traced
body graph rather than by scan outputs. Dynamo also needs to treat `torch.arange`
like other scalar-consuming ops for fake/meta propagation so internal scalar
conversions do not become unbindable pending symbols.

The direct raw scan repro also exposed a separate runtime-assert lowering gap:
`1 / mask.sum().item()` forms `IntTrueDiv(1, u0)` for an unbacked SymInt, and
`PythonReferenceAnalysis` did not implement `int_truediv`. Add the traceable
handler so deferred runtime assertions lower to FX division instead of trying to
construct a SymPy expression with FX proxies.

This keeps the fix scoped to export/fake/runtime-assert tracing behavior. A
broader redesign of every data-dependent `arange` output shape form is outside
this issue.

Fixes #153247
Generated by my agent

Test Plan:
- python test/export/test_export.py TestExport.test_export_arange_data_dependent_float_step TestExport.test_export_scan_data_dependent_arange_float_step TestExport.test_export_direct_scan_empty_init_data_dependent_arange_item_step
- python test/dynamo/test_misc.py MiscTests.test_arange_length_with_float32_dtype MiscTests.test_torch_check MiscTests.test_torch_check_symbolic_shape_rel
- python test/dynamo/test_repros.py -k test_unbacked_arange_in_bounds
- lintrunner -a torch/_refs/__init__.py torch/_higher_order_ops/scan.py torch/_dynamo/variables/torch.py torch/utils/_sympy/reference.py test/export/test_export.py

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the CI failures by marking the raw direct scan regression as an expected failure only for generated strict/strict-serdes/training-decomp variants, where raw torch.ops.higher_order.scan with a Python body is not strict-exportable today. The base non-strict regression remains active, and I resubmitted the ghstack.

**Comment by jansel** [comment]
My agent says I fixed the remaining export CI failure by also marking the generated retraceability strict variant of the raw direct scan regression as an expected failure. The base non-strict regression still runs and passes; the generated strict/retrace variants fail because raw torch.ops.higher_order.scan with a Python body is not supported by those wrappers today. The docs preview failures were from the CI environment missing aws during upload, not from this PR.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
